ssl/statem/extensions.c

```diff
@@ -1444,43 +1444,114 @@ static int final_key_share(SSL_CONNECTION *s, unsigned int context, int sent)
                     && (!s->hit
                         || (s->ext.psk_kex_mode & TLSEXT_KEX_MODE_FLAG_KE_DHE)
                            != 0)) {
-                const uint16_t *pgroups, *clntgroups;
-                size_t num_groups, clnt_num_groups, i;
-                unsigned int group_id = 0;
-
-                /* Check if a shared group exists */
-
-                /* Get the clients list of supported groups. */
-                tls1_get_peer_groups(s, &clntgroups, &clnt_num_groups);
-                tls1_get_supported_groups(s, &pgroups, &num_groups);
-
-                /*
-                 * Find the first group we allow that is also in client's list
-                 */
-                for (i = 0; i < num_groups; i++) {
-                    group_id = pgroups[i];
-
-                    if (check_in_list(s, group_id, clntgroups, clnt_num_groups,
-                                      1)
-                            && tls_group_allowed(s, group_id,
-                                                 SSL_SECOP_CURVE_SUPPORTED)
-                            && tls_valid_group(s, group_id, TLS1_3_VERSION,
-                                               TLS1_3_VERSION, 0, NULL))
-                        break;
+                    const uint16_t *pgroups, *clntgroups;
+                    size_t num_groups, clnt_num_groups, i;
+                    unsigned int group_id = 0;
+
+                    /* Check if a shared group exists */
+
+                    /* Get the clients list of supported groups. */
+                    tls1_get_peer_groups(s, &clntgroups, &clnt_num_groups);
+                    tls1_get_supported_groups(s, &pgroups, &num_groups);
+
+                    /*
+                     * Find the first group we allow that is also in client's list
+                     */
+                    for (i = 0; i < num_groups; i++) {
+                        group_id = pgroups[i];
+
+                        if (check_in_list(s, group_id, clntgroups, clnt_num_groups,
+                                          1)
+                                && tls_group_allowed(s, group_id,
+                                                     SSL_SECOP_CURVE_SUPPORTED)
+                                && tls_valid_group(s, group_id, TLS1_3_VERSION,
+                                                   TLS1_3_VERSION, 0, NULL))
+                            break;
+                    }
+
+                    if (i < num_groups) {
+                        /* A shared group exists so send a HelloRetryRequest */
+                        s->s3.group_id = group_id;
+                        s->hello_retry_request = SSL_HRR_PENDING;
+                        return 1;
+                    }
                 }
+            } else { /* a '!' prefix was used, hence we apply a new group selection methodology */
+                fprintf(stdout, "=====> %d x '!' prefix detected in extensions.c\n", s->ext.s_ext_ks_selection_detected);
```

Comment:
Presumably this goes once this is no longer draft code...
We have a trace API, this could be part of the trace if it should be retained.